p2p/src/codec.rs

```diff
@@ -0,0 +1,268 @@
+    buf.put(peer.id.0.as_bytes());
+}
+
+impl MessageEncoder {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+pub struct MessageDecoder {
+    state: DecodeState,
+}
+
+impl MessageDecoder {
+    pub fn new() -> Self {
+        MessageDecoder {
+            state: DecodeState::MessageType,
+        }
+    }
+}
+
+enum DecodeState {
+    MessageType,
+    Len(u8),
+    Body(u8, usize),
+}
+
+impl Decoder for MessageDecoder {
+    type Item = PeerMessage;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        match self.state {
+            DecodeState::MessageType => {
+                if src.is_empty() {
+                    return Ok(None);
+                }
+                let command_type = src.get_u8();
+                match command_type {
+                    0..=2 => {}
+                    _ => {
+                        return Err(io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            format!("Unknown command: {}", command_type),
+                        ))
+                    }
+                }
+                self.state = DecodeState::Len(command_type);
+                self.decode(src)
+            }
+            DecodeState::Len(m_type) => {
+                if src.len() < 8 {
+                    return Ok(None);
+                }
+                let len = src.get_u64() as usize;
+                self.state = DecodeState::Body(m_type, len);
+                self.decode(src)
+            }
+            DecodeState::Body(m_type, len) => {
+                if src.len() < len {
+                    return Ok(None);
+                }
+                self.state = DecodeState::MessageType;
+                read_message_body(m_type, len, src).map(Some)
+            }
+        }
+    }
+}
+
+fn read_message_body(
+    message_type: u8,
+    len: usize,
+    src: &mut BytesMut,
+) -> Result<PeerMessage, io::Error> {
+    match message_type {
+        0 => {
+            if len != 2 {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("Invalid length for hello message: {}", len),
+                ));
+            }
+            let data = src.get_u16();
+            Ok(PeerMessage::Hello(data))
+        }
+        1 => {
+            let mut peers = vec![];
+            let mut peers_bytes = src.split_to(len);
+            while let Some(peer) = decode_peer_addr(&mut peers_bytes)? {
+                peers.push(peer);
+            }
+            Ok(PeerMessage::Peers(peers))
+        }
+        2 => Ok(PeerMessage::Data(src.split_to(len).to_vec())),
+        m => Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("Unknown message type: {}", m),
+        )),
+    }
+}
+
+fn decode_peer_addr(buf: &mut BytesMut) -> Result<Option<PeerAddr>, io::Error> {
+    if buf.is_empty() {
+        return Ok(None);
```

Comment:
I don't think this is locally correct (but seems to be correct within the scope of the module). This helper is supposed to be running in a context where we are expecting to have all the bytes that we need (like the socket address parsing helpers below), while `Ok(None)` means that we (1) need more bytes and (2) will correctly remember the state of already received bytes. But the whole point of having a big fat length prefix for all message types is that we don't have to remember all intermediate states while waiting for more bytes. Instead, we can just parse them and fail fast. Our messages are quite short and we are not feeding portions of messages to a downstream system, so mandating buffering for the entire message is not a design or performance concern.

Reply:
aha, it's Ok(None) in case we run out of bytes, which is when we finish parsing the list. So i guess, it's okay to leave it as-is.

Reply:
Since it is not clear at first sight, I'll move this check into `read_message_body`.